### PR TITLE
consolidate linting and add it to GitHub Actions

### PR DESCRIPTION
**Pull request by jmhodges**
This consolidates our linting tools into the golangci-lint tool. It also
removes the unsupported, archived golint tool.

Docs about golangci-lint can be found here: https://golangci-lint.run/

Specifics about how to configure specific linters and what they are can
be found at https://golangci-lint.run/usage/linters/

The new GitHub Action integration also will annotate the PR its run on
with the violations it fails instead of just printing them out to the
console. This means we'll get nice lil UI things directly next to the
code in a PR that's causing the linter violation. This is the Action
maintained by the golangci-lint maintainers.

The default configuration of `golangci-lint` includes `errcheck` which
is very handy but was going to cause a significant amount of churn here.
I've made AUT-202 for us to turn it on later.

We use `depguard` to replace the `check-no-crypto11-in-signers` target.
It's a lil funky (I don't like the "**/" prefix that's required and
could cause false positives if we ever gain another `signer` package
name) but it's nice to have it in place for future checks like it. Plus,
it won't have the previous kind of false positives of the old grep.

`staticcheck` is still run in golangci-lint by default and we make sure
it remains enabled. No change there.

Along the way, pick up some new lint errors that the previous linting
missed.

One is that the `main.httpError` func is Sprintf-like, and `go vet`'s
`printf` checks were being violated in a few places. Same deal with some
uses of `fmt.Errorf` without any additional arguments. Some `log.Printf`
calls had the wrong number of arguments for the format string they were
given. One of the `main.httpError` calls is false positive-y and gets a
useless `"%s"` because we have no "zero arguments" function equivalent,
but it's fine.

Our immediate update of the `debug` variable in `main_racing_test.go`
meant the new `ineffassign` linter was unhappy. It's slightly false
positive-y because I suspect it was folks just trying to be clear about
what the zero value of a bool is, but the new code for that is shorter
and more typical Go.

`go vet` ("`govet`" in golangci-lint parlance) also had an "unreachable
code" trigger in the autograph-client codebase. A nice simplification
made it go away.

`golangci-lint` got a 5 minute timeout (`--timeout`) in the GitHub
Actions. That time was chosen arbitrarily, and 5 minutes would be a long
time to wait for a linting check to finish. We could definitely shorten
it and maybe should.

Also, the Makefile gains the ability to install `golangci-lint`, because
I don't have a better pattern for how folks should install it right now.

Updates AUT-158
Updates AUT-197

### CODE REVIEW

File: .github/workflows/tests.yaml
Comment (hneiva) [inline .github/workflows/tests.yaml:69]: nit: do we care if this takes > 5m?

File: .golangci.yaml
Comment (hneiva) [inline .golangci.yaml:7]: Awesome! thanks for leaving a hint :)

**Review by hneiva — APPROVED** [review]
Looks great! Thank you!

File: .github/workflows/tests.yaml
Comment (jmhodges) [reply .github/workflows/tests.yaml:69]: It likely means something is very wrong with our dev environment or our lint tooling. Either would mean we need to address it. The alternative is this taking a long time and folks just going "guess our build is slow" and not noticing til it's too late.

Comment (hneiva) [reply .github/workflows/tests.yaml:69]: heh fair enough!

File: monitor_handler.go
Comment (oskirby) [inline monitor_handler.go:37]: Uhh, I think `go vet` is right in this case. It seems like `httpError()` seems to accept format strings, so if `errstr` contains any percent characters it will expect more args on the call stack.

File: .github/workflows/tests.yaml
Comment (oskirby) [inline .github/workflows/tests.yaml:61]: Rather than `stable` should we instead use the same version specified in the `go.mod`?

File: .circleci/config.yml
Comment (oskirby) [inline .circleci/config.yml:1]: I have a radical suggestion - let's remove the CircleCI linting job entirely if this also introduces a Github version that does the same job. There is limited value in having the same test run in two places and producing the same warning messages.

**Review by oskirby — APPROVED** [review]
I left a couple suggestions - but overall I think this is good.

File: .circleci/config.yml
Comment (hneiva) [reply .circleci/config.yml:1]: oh, I like this one!

File: monitor_handler.go
Comment (jmhodges) [reply monitor_handler.go:37]: I mean, technically I suppose? But the errors are known to not contain format strings (see, for instance, [these lines](https://github.com/mozilla-services/autograph/blob/5172bfb5ebfddc160172926a3d52e1ea4f39bae4/monitor.go#L78-L88)), and if they do, this isn't C: the `fmt.Sprintf` fills them in with `EXTRA` strings. See https://pkg.go.dev/fmt#hdr-Format_errors for how Go handles format string errors

The alternative would be to write another httpError that just didn't take extra args at all and that didn't seem worth it.

File: .github/workflows/tests.yaml
Comment (jmhodges) [reply .github/workflows/tests.yaml:61]: I was hoping to not have another spot where we're updating the Go version across our codebase. The Go that runs the lint tooling being different than what we build the code seemed fine, given the [Go backwards compatibility](https://go.dev/blog/compat) guarantees.

If you're interested in this, I'm down, but it seemed like future noise and something we could lock down later if we truly needed to.

File: .circleci/config.yml
Comment (jmhodges) [reply .circleci/config.yml:1]: I was going to post a follow-up PR to remove the CircleCI job after this was merged and after I'd removed the CircleCI linter job from the branch protection configuration.

So, if I add this new Linting check to the branch protection requirements, that means everyone has to wait for this PR to merge before they can get their PRs will pass. Similarly, if I remove the CircleCI lint checks before this is merged in, we lose the linting til then. 

I'll admit I prefer to practice some amount of "do the backwards compatible thing, then do the change after".

File: .github/workflows/tests.yaml
Comment (jmhodges) [reply .github/workflows/tests.yaml:61]: Like, the violation pops up is actually because of the `Sprintf` call, but it doesn't go away when I do this work conditionally (because Sprintf doesn't actually care much). They're just pattern matching on "string + args". See this patch:

```
diff --git a/errors.go b/errors.go
index 9818b5b1..ba636d0b 100644
--- a/errors.go
+++ b/errors.go
@@ -23,7 +23,13 @@ func httpError(w http.ResponseWriter, r *http.Request, errorCode int, errorMessa
 		"code": errorCode,
 		"rid":  rid,
 	}).Errorf(errorMessage, args...)
-	msg := fmt.Sprintf(errorMessage, args...)
+	var msg string
+	if len(args) == 0 {
+		msg = errorMessage
+	} else {
+		msg = fmt.Sprintf(errorMessage, args...)
+	}
+
 	msg += "\r\nrequest-id: " + rid
 	// when nginx is in front of go, nginx requires that the entire
 	// request body is read before writing a response.
diff --git a/monitor_handler.go b/monitor_handler.go
index da97be8c..745ce17f 100644
--- a/monitor_handler.go
+++ b/monitor_handler.go
@@ -34,8 +34,7 @@ func (m *monitor) handleMonitor(w http.ResponseWriter, r *http.Request) {
 
 	for _, errstr := range m.sigerrstrs {
 		if errstr != "" {
-			// "%s" is a go vet false positive correction
-			httpError(w, r, http.StatusInternalServerError, "%s", errstr)
+			httpError(w, r, http.StatusInternalServerError, errstr)
 			return
 		}
 	}
```

File: monitor_handler.go
Comment (jmhodges) [reply monitor_handler.go:37]: Like, the violation that pops up is actually because of the `Sprintf` call, but it doesn't go away when I do this work conditionally (because Sprintf doesn't actually care much). The linter is just pattern matching on "string + args in function signature". See this patch and how it doesn't make the lint violation go away:

```patch
diff --git a/errors.go b/errors.go
index 9818b5b1..ba636d0b 100644
--- a/errors.go
+++ b/errors.go
@@ -23,7 +23,13 @@ func httpError(w http.ResponseWriter, r *http.Request, errorCode int, errorMessa
 		"code": errorCode,
 		"rid":  rid,
 	}).Errorf(errorMessage, args...)
-	msg := fmt.Sprintf(errorMessage, args...)
+	var msg string
+	if len(args) == 0 {
+		msg = errorMessage
+	} else {
+		msg = fmt.Sprintf(errorMessage, args...)
+	}
+
 	msg += "\r\nrequest-id: " + rid
 	// when nginx is in front of go, nginx requires that the entire
 	// request body is read before writing a response.
diff --git a/monitor_handler.go b/monitor_handler.go
index da97be8c..745ce17f 100644
--- a/monitor_handler.go
+++ b/monitor_handler.go
@@ -34,8 +34,7 @@ func (m *monitor) handleMonitor(w http.ResponseWriter, r *http.Request) {
 
 	for _, errstr := range m.sigerrstrs {
 		if errstr != "" {
-			// "%s" is a go vet false positive correction
-			httpError(w, r, http.StatusInternalServerError, "%s", errstr)
+			httpError(w, r, http.StatusInternalServerError, errstr)
 			return
 		}
 	}
```

Comment (jmhodges) [reply monitor_handler.go:37]: I guess maybe you're saying you just want the comment deleted?

File: .circleci/config.yml
Comment (oskirby) [reply .circleci/config.yml:1]: I am okay with doing the work in two PRs.

File: monitor_handler.go
Comment (oskirby) [reply monitor_handler.go:37]: Yeah, that's all I was trying to say - just remove the comment and it's fine.

Comment (jmhodges) [reply monitor_handler.go:37]: *embarrassed face* can do

File: .github/workflows/tests.yaml
Comment (oskirby) [reply .github/workflows/tests.yaml:61]: I think the core of my concern is the hypothetical scenario:
1. We develop our code against Go version X and the linter is very happy with our code.
2. Go releases a new version, X+1 as the new stable, and introduces a whole bunch of lints.
3. The golangci linter uses the stable version rather than our `go.mod` when deciding which lints apply.
4. Our CI now goes red even though we changed nothing in our code.

If the linter workflow examines our `go.mod` to determine the applicable lints, then I am fine leaving this PR as-is.

Comment (jmhodges) [reply .github/workflows/tests.yaml:61]: Understood. I believe go.mod is looked at for some of the lints in `go vet` (e.g. the loop varialbe related ones). I'm less worried about breaking the CI cuz it's easy enough to plot down a specific Go version if it turns out we need it and like picking up new, better lints without having to work at it.

But I'm happy to adjust to my reviewer here.

Comment (jmhodges) [reply .github/workflows/tests.yaml:61]: oh jeez, I had forgotten `go-version-file: 'path/to/go.mod'` existed. Just using that is clearly better. This is perhaps what you meant?

**Pull request merged**
